### PR TITLE
Added Contour as an alternative for nginx ingress

### LINKED ISSUE #22: Contour
https://projectcontour.io/ as an alternative to nginx.

### PR DESCRIPTION
**Pull request by K-DOT**
Fixes #22

### CODE REVIEW

**Comment by moshloop** [comment]
Can you common on the PR: "Fixes #22" so that when this PR merges, the issue automatically gets closed ?

Can we also rather vendor the YAML into manifests rather than using the helm chart ? 
https://raw.githubusercontent.com/projectcontour/contour/release-1.1/examples/render/contour.yaml

**Comment by K-DOT** [comment]
> Can you common on the PR: "Fixes #22" so that when this PR merges, the issue automatically gets closed ?
> 
> Can we also rather vendor the YAML into manifests rather than using the helm chart ?
> https://raw.githubusercontent.com/projectcontour/contour/release-1.1/examples/render/contour.yaml

Sure

File: pkg/phases/base/install.go [hunk 1]
```diff
@@ -95,11 +96,16 @@ func Install(platform *platform.Platform) error {
...
 	}
 
-	if platform.Nginx == nil || !platform.Nginx.Disabled {
+	if platform.Nginx != nil && !platform.Nginx.Disabled {
 		log.Infof("Installing Nginx Ingress Controller")
 		if err := platform.ApplySpecs("", "nginx.yml"); err != nil {
 			log.Errorf("Error deploying nginx: %s\n", err)
 		}
+	} else if platform.Contour != nil && !platform.Contour.Disabled {
+		log.Infof("Skipping deployment of nginx. Contour will be deployed")
```
Comment (moshloop) [inline pkg/phases/base/install.go]:
```suggestion
		log.Infof("Installing Contour Ingress Controller")
```

File: pkg/phases/base/install.go [hunk 2]
```diff
@@ -95,11 +96,16 @@ func Install(platform *platform.Platform) error {
 		}
 	}
 
-	if platform.Nginx == nil || !platform.Nginx.Disabled {
+	if platform.Nginx != nil && !platform.Nginx.Disabled {
```
Comment (moshloop) [inline pkg/phases/base/install.go]:
```suggestion
        // Deploy nginx ingress controller if neither nginx or contour is specified
	if (platform.Nginx == nil && platform.Contour == nil) || (platform.Nginx != nil && !platform.Nginx.Disabled) {
```

Comment (moshloop) [reply pkg/phases/base/install.go]: Perhaps wrap the if statement into a `platform.GetIngressController` for reuse

File: pkg/phases/base/test.go [hunk 3]
```diff
@@ -19,12 +20,12 @@ func Test(platform *platform.Platform, test *console.TestResults) {
...
-	if platform.Nginx == nil || !platform.Nginx.Disabled {
+	if (platform.Nginx == nil || !platform.Nginx.Disabled) && (platform.Contour == nil || platform.Contour.Disabled )  {
 		k8s.TestNamespace(client, "ingress-nginx", test)
 	}
 
 	if platform.Minio == nil || !platform.Minio.Disabled {
 		k8s.TestNamespace(client, "minio", test)
 	}
-
+	contour.Test(platform, test)
```
Comment (moshloop) [inline pkg/phases/base/test.go]: this should be tested inline e.g.

```golang
if platform.GetIngressController() == "contour") {
    k8s.TestNamespace(client, "contour")
} else {
    k8s.TestNamespace(client, "nginx-ingress")
}
```
The tests would only be considered a skip if both nginx and contour are disabled.

File: test/common.yml [hunk 4]
```diff
@@ -80,3 +79,6 @@ fluentd-operator:
   repository: jvassev/kube-fluentd-operator
 eck:
   version: 1.0.0
+contour:
+  version: v0.15.0
+  chartVersion: 0.2.0
```
Comment (moshloop) [inline test/common.yml]:
```suggestion
```

File: test/e2e.sh [hunk 5]
```diff
@@ -43,41 +43,41 @@ $BIN deploy stubs -v
 
 $BIN test base --wait 200
 
-$BIN deploy pgo install -v
+# $BIN deploy pgo install -v
```
Comment (moshloop) [inline test/e2e.sh]: I think these were inadvertently committed

File: go.mod [hunk 6]
```diff
@@ -6,13 +6,15 @@ require (
 	cloud.google.com/go v0.47.0 // indirect
 	cloud.google.com/go/bigquery v1.2.0 // indirect
 	cloud.google.com/go/storage v1.2.1 // indirect
+	github.com/alecthomas/kingpin v2.2.6+incompatible
```
Comment (moshloop) [inline go.mod]: Don't bump module versions unnecessarily - if they need to be bumped, it is normally better todo it in a new PR

**Comment by moshloop** [comment]
It looks like you may have made some of these commits with a different git config user.name or user.email which is why the CLA is failing.

File: go.mod (hunk 6 above)
Comment (K-DOT) [reply go.mod]: This new dependency was added automatically. I don't know where is it used.

**Comment by K-DOT** [comment]
> It looks like you may have made some of these commits with a different git config user.name or user.email which is why the CLA is failing.
Yes, I changed my laptop and it was set as default. Can we fix it?

**Comment by teodor-pripoae** [comment]
@K-DOT  You can rebase with master and create one new commit instead of the 5.

**Comment by teodor-pripoae** [comment]
It will fix the email

**Pull request closed without merging**